File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string st) {
        if(st.size() == 0) {
            return 0;
        }
        
        int start = 0;
        int end = 0;
        int maxi = 0;
        int len = st.size();
        
        char ch;
        unordered_set<char> s;
        
        while(end < len) {
            ch = st[end];
            
            while(s.find(ch) != s.end()) {
                s.erase(st[start]);
                start++;
            }
            
            s.insert(ch);
            maxi = max(maxi , end - start+1);
            end++;
        }
        
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s == null || s.length() == 0) {
            return 0;
        }
        
        int len = s.length();
        int start = 0;
        int end = 0;
        int maxlength = 0;
        char ch;
        
        Set<Character> uniquechar = new HashSet<>();
        
        while(end < len) {
            ch = s.charAt(end);
            
            while(uniquechar.contains(ch)) {
                uniquechar.remove(s.charAt(start));
                start++;
            }
            
            uniquechar.add(ch);
            maxlength = Math.max(maxlength , end - start+1);
            end++;
        }
        
        return maxlength;
    }
}